### ISSUE TITLE
False positive unreachable error when overload + NoReturn is used

### ISSUE THREAD

**Issue by kasium** [pylint-dev/pylint#10785]
### Bug description

Below code should be valid but pylint finds an issue. It seems to be caused by the NoReturn

```python
# x.py
from y import create_client as mart_create_client
client = mart_create_client(version=2)
x = 1

# y.py
from typing import Literal, NoReturn, overload

@overload
def create_client(version: int = ...) -> NoReturn: ...

@overload
def create_client(version: Literal[2] = ...) -> int: ...

def create_client(version: int = 2) -> int:
    return 1
```

### Command used

```shell
pylint x.py
```

### Pylint output

```python
x.py(6): [W0101(unreachable)] Unreachable code
```

### Expected behavior

no error

### Pylint version

```shell
pylint 4.0.4
astroid 4.0.2
Python 3.14.0 (main, Nov 10 2025, 09:33:32) [GCC 7.5.0]
```

### OS / Environment

SUSE Linux

**Comment by shunping** [comment]
I observed the same false positive when using pylint with hypothesis>6.148.3.

The regression appears to be triggered by a [change](https://github.com/HypothesisWorks/hypothesis/commit/af293bca516bbdf4a498cf1c345989555f74db30#diff-44aaab5d594040b18d0e6340e5e809d978f0f8c2ebf77657a31c9f849c4db1dfR53) in hypothesis 6.148.4 from PR https://github.com/HypothesisWorks/hypothesis/pull/4611 that added an `@overload` returning `NoReturn` to the `assume` function.
 

## Code to reproduce:
```python
"""A simple test to show pylint false positives on hypothesis>6.148.3"""

import unittest

from hypothesis import assume, given, settings
from hypothesis import strategies as st


def calculate_percentage(x, y):
    """The function to test"""
    return (x / y) * 100

class SimpleTest(unittest.TestCase):
    """Test class"""
    @settings(max_examples=10, deadline=None)
    @given(x=st.floats(min_value=1, max_value=100), y=st.floats(min_value=1, max_value=100))
    def test_percentage(self, x, y):
        """Tests that the percentage calculation holds true."""
        assume(x <= y)
        result = calculate_percentage(x, y)
        print(x, y)

        self.assertLessEqual(result, 100.0)


if __name__ == '__main__':
    unittest.main()
```

## Pylint Output
```
************* Module run
run.py:20:8: W0101: Unreachable code (unreachable)
```

## Pylint Version
```
pylint==4.0.4
astroid==4.0.2
Python 3.12.10
```

## OS / Environment
```
macOS
```

**Comment by shunping** [comment]
I ran a bisect on the versions of pylint and found that `pylint 3.2.4` with `astroid==3.2.4` was the first version where such false positive appeared.

The change log did include a false positive fix related to `NoReturn`:
- https://pylint.readthedocs.io/en/latest/whatsnew/3/3.2/index.html#what-s-new-in-pylint-3-2-4
- https://github.com/pylint-dev/pylint/issues/9674
- https://github.com/pylint-dev/pylint/pull/9714

**Comment by CodePapayas** [comment]
Hi, can I be assigned this issue? I would like to attempt to resolve it. Thank you!

**Comment by Pierre-Sassoulas** [comment]
Sure, done !

**Comment by shunping** [comment]
Thank you for the fix! @CodePapayas
